src/main/java/build/buildfarm/common/resources/ResourceParser.java

```diff
@@ -0,0 +1,201 @@
+   */
+  public static UploadBlobRequest parseUploadBlobRequest(String resourceName) {
+    // Find the index of the keyword.  This will give us an initial index to extract information
+    // from.
+    String[] segments = tokenize(resourceName);
+    MutableInt index = startIndex(segments, Resource.TypeCase.UPLOAD_BLOB_REQUEST);
+
+    // Extract each of the following segments:
+    // `{instance_name}/uploads/{uuid}/blobs/{hash}/{size}{/optional_metadata}` or
+    // `{instance_name}/uploads/{uuid}/compressed-blobs/{compressor}/{uncompressed_hash}/{uncompressed_size}{/optional_metadata}`
+    UploadBlobRequest.Builder builder = UploadBlobRequest.newBuilder();
+    builder.setInstanceName(
+        String.join("/", Arrays.copyOfRange(segments, 0, index.getAndIncrement())));
+    builder.setUuid(segments[index.getAndIncrement()]);
+    builder.setBlob(parseBlobInformation(segments, index));
+    if (index.intValue() < segments.length) {
+      builder.setMetadata(
+          String.join("/", Arrays.copyOfRange(segments, index.intValue(), segments.length)));
+    }
+    return builder.build();
+  }
+  /**
+   * @brief Parse the resource name into a specific type.
+   * @details Assumes the URI is already is already in the download blob format.
+   * @param resourceName The resource name URI defined by REAPI.
+   * @return The resource type derived from the URI.
+   * @note Suggested return identifier: request.
+   */
+  public static DownloadBlobRequest parseDownloadBlobRequest(String resourceName) {
+    // Find the index of the keyword.  This will give us an initial index to extract information
+    // from.
+    String[] segments = tokenize(resourceName);
+    MutableInt index = startIndex(segments, Resource.TypeCase.DOWNLOAD_BLOB_REQUEST);
+
+    // Extract each of the following segments: `{instance_name}/blobs/{hash}/{size}` or
+    // `{instance_name}/compressed-blobs/{compressor}/{uncompressed_hash}/{uncompressed_size}`
+    DownloadBlobRequest.Builder builder = DownloadBlobRequest.newBuilder();
+    builder.setInstanceName(String.join("/", Arrays.copyOfRange(segments, 0, index.intValue())));
+    builder.setBlob(parseBlobInformation(segments, index));
+    return builder.build();
+  }
+  /**
+   * @brief A mapping of REAPI keywords to their respective resource types.
+   * @details This lookup table can be used for parsing.
+   * @return A lookup table for keyword to resource type.
+   * @note Suggested return identifier: keywordMap.
+   */
+  private static HashMap<String, Resource.TypeCase> keywordResourceMap() {
+    HashMap<String, Resource.TypeCase> keywords = new HashMap<>();
+    keywords.put("blobs", Resource.TypeCase.DOWNLOAD_BLOB_REQUEST);
+    keywords.put("compressed-blobs", Resource.TypeCase.DOWNLOAD_BLOB_REQUEST);
+    keywords.put("uploads", Resource.TypeCase.UPLOAD_BLOB_REQUEST);
+    keywords.put("operations", Resource.TypeCase.STREAM_OPERATION_REQUEST);
+    return keywords;
```

Comment:
Recommend `ImmutableMap.of` here, rather than the hash map construction, and make this into a publicly accessible static property (ImmutableMap will guarantee attr_reader) that exists as a single instance.

Reply:
fixed